src/cluster_asm.c

```diff
@@ -365,6 +369,161 @@ void asmTaskFree(asmTask *task) {
+
+    /* Add slot ranges sds */
+    sds slot_ranges_str = slotRangeArrayToString(task->slot_ranges);
+    serialized = sdscatprintf(serialized, "%s", slot_ranges_str);
+    sdsfree(slot_ranges_str);
+
+    return serialized;
+}
+
+/* Deserialize ASM task information from a string and create a complete asmTask.
+ * Format: "task_id:source_node:dest_node:operation:state:slot_ranges"
+ * Returns a new asmTask on success, NULL on failure. */
+asmTask *asmTaskDeserialize(sds data) {
+    int count, idx = 0;
+    asmTask *task = NULL;
+    if (!data || sdslen(data) == 0) return NULL;
+
+    sds *parts = sdssplitlen(data, sdslen(data), ":", 1, &count);
+    if (count < 6) goto err;
+
+    /* Parse task ID */
+    if (sdslen(parts[idx]) == 0) goto err;
+    task = asmTaskCreate(parts[idx]);
+    if (!task) goto err;
+    idx++;
+
+    /* Parse source node ID */
+    if (sdslen(parts[idx]) != CLUSTER_NAMELEN) goto err;
+    memcpy(task->source, parts[idx], CLUSTER_NAMELEN);
+    idx++;
+
+    /* Parse destination node ID */
+    if (sdslen(parts[idx]) != CLUSTER_NAMELEN) goto err;
+    memcpy(task->dest, parts[idx], CLUSTER_NAMELEN);
+    idx++;
+
+    /* Parse operation type */
+    if (!strcasecmp(parts[idx], "import")) {
+        task->operation = ASM_IMPORT;
+    } else if (!strcasecmp(parts[idx], "migrate")) {
+        task->operation = ASM_MIGRATE;
+    } else {
+        goto err;
+    }
+    idx++;
+
+    /* Parse state */
+    task->state = ASM_NONE; /* Default state */
+    for (int state = ASM_NONE; state <= ASM_RDBCHANNEL_TRANSFER; state++) {
+        if (!strcasecmp(parts[idx], asmTaskStateToString(state))) {
+            task->state = state;
+            break;
+        }
+    }
+    idx++;
+
+    /* Parse slot ranges */
+    task->slot_ranges = slotRangeArrayFromString(parts[idx]);
+    if (!task->slot_ranges) goto err;
+    idx++;
+
+    sdsfreesplitres(parts, count);
+    return task;
+
+err:
+    if (task) asmTaskFree(task);
+    sdsfreesplitres(parts, count);
+    return NULL;
+}
+
+/* Notify slaves about ASM task information to maintain consistency during slot migration.
+ * This function sends a CLUSTER SYNCSLOTS CONF ASM-TASK command to all connected slaves
+ * with the serialized task information. */
+void asmNotifySlavesStateChange(struct asmTask *task) {
+    if (!server.cluster_enabled || !clusterNodeIsMaster(getMyClusterNode())) return;
+
+    /* Do not propagate migrate task to slaves, as slaves never migrate data. */
+    if (task->operation == ASM_MIGRATE) return;
+
+    /* Create command arguments for CLUSTER SYNCSLOTS CONF ASM-TASK */
+    robj *argv[5];
+    argv[0] = createStringObject("CLUSTER", 7);
+    argv[1] = createStringObject("SYNCSLOTS", 9);
+    argv[2] = createStringObject("CONF", 4);
+    argv[3] = createStringObject("ASM-TASK", 8);
+    argv[4] = createObject(OBJ_STRING, asmTaskSerialize(task));
+
+    /* Send the command to all slaves */
+    replicationFeedSlaves(server.slaves, -1, argv, 5);
```

Comment:
I wonder if we need to check pause actions before writing the backlog. For trim, we moved this logic to beforeSleep() but for this notification, it seems a lot of work.

Reply:
good point.

I checked code where calling `asmNotifyStateChange`, i think it should be okay. The propagate doesn't like active trim that last a long time.

for STARTED, pausing must be over, otherwise, we can't start a new task. and during import, if we want to pause server, we will cancel the task and propagate this notification first before calling pauseActions

for migrate done, we pause the server first, we already unpause server in clusterAsmOnEvent before propagating.

 besides, it should not cause much damage even if server is pause, i mean redis won't crash, maybe some operations may failed or timeout

Reply:
okay, thanks. If it does not crash, we should be good.